Client/Frontend/Brave Notifications/BraveNotificationsPresenter.swift

```diff
@@ -0,0 +1,266 @@
+        } else {
+          guard let presentingController = self.parent else { return }
+          self.display(notification: self.notificationsQueue.popLast()!, from: presentingController)
+        }
+      }
+    }
+  }
+  
+  private func enqueueNotification(_ notification: BraveNotification) {
+    // We will skip duplication checking for notifications that have empty id. These notifications are usually custom ads
+    if !notification.id.isEmpty,
+       notificationsQueue.contains(where: { $0.id == notification.id }) {
+      return
+    }
+    
+    if let index = notificationsQueue.firstIndex(where: { $0.priority < notification.priority }) {
+      notificationsQueue.insert(notification, at: index + 1)
+    } else {
+      notificationsQueue.insert(notification, at: 0)
+    }
+  }
+  
+  deinit {
+    dismissTimers.forEach({ $0.value.invalidate() })
+  }
+  
+  // MARK: - Actions
+  
+  private var dismissTimers: [String: Timer] = [:]
+  
+  private func setupTimeoutTimer(for notification: BraveNotification, interval: TimeInterval) {
+    if let timer = dismissTimers[notification.id] {
+      // Invalidate and reschedule
+      timer.invalidate()
+    }
+    dismissTimers[notification.id] = Timer.scheduledTimer(withTimeInterval: interval, repeats: false, block: { [weak self] _ in
+      guard let self = self else { return }
+      self.hide(notification)
+      notification.willDismiss(timedOut: true)
+    })
+  }
+  
+  // Distance travelled after decelerating to zero velocity at a constant rate
+  func project(initialVelocity: CGFloat, decelerationRate: CGFloat) -> CGFloat {
+    return (initialVelocity / 1000.0) * decelerationRate / (1.0 - decelerationRate)
+  }
+  
+  private var panState: CGPoint = .zero
+  @objc private func dismissPannedAdView(_ pan: UIPanGestureRecognizer) {
+    guard let notificationView = pan.view else { return }
+    
+    switch pan.state {
+    case .began:
+      panState = notificationView.center
+      // must be the current visible notification
+      guard let notification = visibleNotification, notification.view == notificationView else {
+        return
+      }
+      // Make sure to stop the dismiss timer
+      dismissTimers[notification.id]?.invalidate()
+    case .changed:
+      notificationView.transform.ty = min(0, pan.translation(in: notificationView).y)
+    case .ended:
+      let velocity = pan.velocity(in: notificationView).y
+      let y = min(panState.y, panState.y + pan.translation(in: notificationView).y)
+      let projected = project(initialVelocity: velocity, decelerationRate: UIScrollView.DecelerationRate.normal.rawValue)
+      if y + projected < 0 {
+        // must be the current visible notification
+        guard let notification = visibleNotification, notification.view == notificationView else {
+          return
+        }
+        hide(notificationView: notificationView, velocity: velocity)
+        notification.willDismiss(timedOut: false)
```

Comment:
`willDismiss` should be called _before_ calling hide, otherwise name it `didDismiss`